feat(native-filters): Re-arrange controls in FilterBar

kgabryje opened the pull request:
### SUMMARY
This PR implements a part of Stage 1 of native filters rework.
Changes included:
- Replace a pencil icon with text button - "+ Add/Edit Filters"
- Move "apply filters" and "clear all" buttons to the bottom of filter bar and make them sticky. The background of the buttons container is semi-transparent (with a gradient) so that scrolling filters looks smoother
- Adjust the text in empty state

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


https://user-images.githubusercontent.com/15073128/154497849-c549f8d2-8138-4c60-a4c8-f128f9362f23.mov

<img width="263" alt="image" src="https://user-images.githubusercontent.com/15073128/154497953-552ab0f6-4363-47ae-8cc3-03a8c0a2edf6.png">

### TESTING INSTRUCTIONS
Make sure that every button in Filters Bar works exactly like in the old version

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

kgabryje committed:
feat(native-filters): Re-arrange controls in FilterBar

kgabryje committed:
Typo fix

kgabryje committed:
Add tests

kgabryje committed:
Fix tests

michael-s-molina commented:
@kgabryje Did you test it with the white background instead of the semi-transparent one? For me, when a filter is passing behind the buttons it looks really weird like the layout broke or something.

michael-s-molina commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/ActionButtons.test.tsx:
```diff
@@ -0,0 +1,85 @@
+  expect(screen.getByText('Apply filters').parentElement).toBeEnabled();
+});
+
+test('should render the "Clear all" button as disabled', () => {
```
Since we're not passing any additional prop to the default render here, I think this test is sufficient and we could remove the one named "should render the "Clear all" option". Because this test is already checking that the "Clear all" button exists and is also disabled.

michael-s-molina commented on superset-frontend/src/components/Button/index.tsx:
```diff
@@ -200,14 +200,16 @@ export default function Button(props: ButtonProps) {
           marginLeft: theme.gridUnit * 2,
         },
-        '& :first-of-type': {
+        '& > :first-of-type': {
```
Can you explain why this was needed? I'm just trying to understand if this change can impact other buttons in the application.

kgabryje replied:
This selector was targeting every "first of type" element being a descendant of `Button`. In this case (see screenshot below), it would target both `<span>` and `<svg>` elements, so in result we'd apply margin twice. With this change, we'll only target direct children of `Button`.

<img width="549" alt="image" src="https://user-images.githubusercontent.com/15073128/154531575-e15b41c3-4c69-4c0a-877f-7f3f38a4c527.png">

michael-s-molina replied:
Understood. I just don't know if this change can cause a regression elsewhere.

michael-s-molina commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/index.tsx:
```diff
@@ -0,0 +1,109 @@
+  dataMaskSelected,
+  isApplyDisabled,
+}: ActionButtonsProps) => {
+  const isClearAllDisabled = useMemo(
```
Is it possible to invert the logic and check if one filter has an applied value? That way we don't need to loop through all filters.

michael-s-molina commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/FilterControls/FilterControls.tsx:
```diff
@@ -42,6 +42,9 @@ import { buildCascadeFiltersTree } from './utils';
 
 const Wrapper = styled.div`
   padding: ${({ theme }) => theme.gridUnit * 4}px;
+  // 108px padding to make room for buttons with position: absolute
```
If we change to a white background then could move the bottom buttons outside the scrollable area.

michael-s-molina commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Header/index.tsx:
```diff
@@ -126,29 +87,18 @@ const Header: FC<HeaderProps> = ({
+            dashboardId={dashboardId}
+            createNewOnOpen={filterValues.length === 0}
+          >
+            <>
```
I noticed that the plus icon is not vertically aligned with the text. Can we fix it? Is the fragment necessary here?

michael-s-molina commented:
> > @kgabryje Did you test it with the white background instead of the semi-transparent one? For me, when a filter is passing behind the buttons it looks really weird like the layout broke or something.
> > <img alt="Screen Shot 2022-02-17 at 1 44 49 PM" width="268" src="https://user-images.githubusercontent.com/70410625/154529273-478c1535-e300-40e9-a52e-8f8af069ef5e.png">
> 
> That change was discussed with @kasiazjc - the purpose is to indicate that there is hidden content under the buttons and user can scroll down.

But then when there's available content to the top wouldn’t be the same?

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Header/index.tsx:
```diff
@@ -126,29 +87,18 @@ const Header: FC<HeaderProps> = ({
+            createNewOnOpen={filterValues.length === 0}
+          >
+            <>
+              <Icons.PlusSmall /> Add/Edit Filters
```
👍

kasiazjc commented:
> > > That change was discussed with @kasiazjc - the purpose is to indicate that there is hidden content under the buttons and user can scroll down.
> > 
> > 
> > But then when there's available content to the top wouldn’t be the same?
> 
> One thing we could do is to keep the scroll always visible when there is one or add some icon like arrows top/bottom or 3 dots to indicate that there is more content available scrolling top or bottom. I prefer to keep the scroll always visible.

@michael-s-molina Placement of the buttons at the top and bottom are two separate issues/flows. You scroll from the top to bottom, so when you scroll down you probably remember that there is something at the top. It may not be on the other hand intuitive that you can scroll down especially when the content/field hidden behind the buttons at the bottom is not partially visible. 

From the design perspective the way we have scroll bar implemented right now is the intuitive/universal option and looks much cleaner (also that's how it's implemented in browsers etc.). Also adding the gradient behind the bottom section, especially when it comes to the buttons, helps to make the design lighter and more intuitive without it being to blocky - another white block at the bottom, especially when user's screen is smaller, can clutter the page (we already have a lot of different sections, panels etc.). We can change the % of the gradient, but I think this solutions works best here.

michael-s-molina commented:
@kasiazjc Good point about the different flow. I also don't like additional icons, I just thought that showing the scroll was sufficient to indicate that there are hidden items, and it works both ways. But no problem, it was just an impression that I had while testing. Thanks for the explanation 😉

kasiazjc commented:
@michael-s-molina and I appreciate the discussion 🤠 🤝  Thanks!

jinghua-qa commented:
/testenv up

kgabryje committed:
Address PR reviews

kgabryje committed:
Add 1px bottom padding to icon

kgabryje committed:
Fix test

kgabryje committed:
Tiny refactor

michael-s-molina reviewed (approved):
Code LGTM. @jinghua-qa Can we get QA's approval?

geido commented:
Hey @kgabryje some points / thoughts after manual testing it.

- When scrolling the filter bar changes height and the button is also changing its vertical position

https://user-images.githubusercontent.com/60598000/154705825-1bf9f078-9b24-42f3-b126-4c0b25cee897.mp4

- Don't want to oppose anything but as a personal opinion, I also agree with @michael-s-molina that the transparent bottom actions on scroll look weird. It required a second look to understand it was a design choice. 

<img width="1177" alt="Screen Shot 2022-02-18 at 16 56 53" src="https://user-images.githubusercontent.com/60598000/154707917-a7735284-58e8-444e-81f0-797b8cd6a724.png">

jinghua-qa commented:
> Code LGTM. @jinghua-qa Can we get QA's approval?

I will test it today

kgabryje committed:
Change buttons background

jinghua-qa commented:
I have 2 questions for the new look:
1, the clear all button is always grey out and make me feel like it is disable, also i feel like the buttons are not consistent, for example the clear button did not have a background but apply all have a grey back ground, and it makes me feel like apply is more standout

<img width="1792" alt="Screen Shot 2022-02-18 at 12 38 20 PM" src="https://user-images.githubusercontent.com/81597121/154758060-2cf8eb2a-3abc-4846-ac9d-f56e7f232abc.png">

2, if i have multiple filters, the scroll bar is hidden, and user need to move the mouse into the filters area to be able to scroll, and there in no clear window to guide the user to move the mouse, which could be a little confusing?

https://user-images.githubusercontent.com/81597121/154758393-60d9b221-f5dc-4728-8f0d-b048c3b5b739.mov

kgabryje committed:
Add hover state to Clear all button

kgabryje committed:
Fix Clear All button logic so it clears all selections

kgabryje commented:
Thanks for the comments @jinghua-qa! After consulting with @kasiazjc I changed the default color of "Clear all" to a darker one (first screenshot) and changed the color on hover (second screenshot). The light grey color is now only used when button is disabled. As for the apply button standing out - it does stand out by design as it calls a more important action than clear all.
<img width="211" alt="image" src="https://user-images.githubusercontent.com/15073128/154951951-8aa59d5d-eb90-407a-a624-b6a442a0d0ae.png">
<img width="176" alt="image" src="https://user-images.githubusercontent.com/15073128/154951991-083ddd5c-8240-4bdd-922f-144612e180ce.png">

kasiazjc commented:
> I have 2 questions for the new look: 1, the clear all button is always grey out and make me feel like it is disable, also i feel like the buttons are not consistent, for example the clear button did not have a background but apply all have a grey back ground, and it makes me feel like apply is more standout
> 
> <img alt="Screen Shot 2022-02-18 at 12 38 20 PM" width="1792" src="https://user-images.githubusercontent.com/81597121/154758060-2cf8eb2a-3abc-4846-ac9d-f56e7f232abc.png">
> 
> 2, if i have multiple filters, the scroll bar is hidden, and user need to move the mouse into the filters area to be able to scroll, and there in no clear window to guide the user to move the mouse, which could be a little confusing?
> 
>  filter.bar.in.visible.mov

Thank you for feedback @jinghua-qa!

1. We want to make the apply all more visible than the clear all, so that's why we have this hierarchy + difference. Plus, it's a standard pattern for this kind of components - usually in filter components "Clear all" action is less visible and shown as a link, so I wouldn't say it's inconsistent when it comes to types of the buttons, it's just a different pattern. I agree though, that clear all seemed like it's disabled at all times, so @kgabryje added different states for disabled, active and hover, but the idea is to still keep this hierarchy we have and not overuse the colours. 
2. I think the hidden scroll bar shouldn't be a problem as it's a natural behaviour/flow. Design wise, scroll bars are usually hidden also in the components that have an internal scroll, not only websites or apps. The gradient that we added should be able to help with it a little bit. It works the same in different parts of app though and in the filter boxes, tables etc.

kgabryje commented:
@geido @michael-s-molina I managed to fix the issue with jumping bottom buttons and filter bar container. I used `position: fixed` on the buttons container, which means now they are detached from filter bar. Filter bar on the other hand is still jumpy... but user doesn't see it because the bottom part is covered by buttons 😄

https://user-images.githubusercontent.com/15073128/154964192-38126d5f-77a7-479a-8f86-835405b509ab.mov

Of course, we can keep looking for a solution that actually prevents the filter bar from jumping, but I think it's a much less pressing issue now.

geido commented:
/testenv up

jinghua-qa commented:
@kgabryje @kasiazjc Thank you for the explanation and I think the new fix for the apply/clear button look better.

I found another cosmetic issue of scrolling up the select box, i think sometimes if user scroll up the bar when select box is open, the select box will be scroll up too, also the content of dashboard is scrolled up

https://user-images.githubusercontent.com/81597121/155091891-8e15f55a-80ca-479a-998a-ae6f254614d7.mov

kgabryje commented:
> @kgabryje @kasiazjc Thank you for the explanation and I think the new fix for the apply/clear button look better.
> 
> I found another cosmetic issue of scrolling up the select box, i think sometimes if user scroll up the bar when select box is open, the select box will be scroll up too, also the content of dashboard is scrolled up
> 
>  scrolling.issue.mov

That's super weird 😕 I can't reproduce this issue locally, but I'm fairly sure that it's not related to my changes. @geido @michael-s-molina Are you guys able to reproduce?

kgabryje committed:
Remove redundant useEffect

kgabryje committed:
Set zindex higher than loading icon

kgabryje committed:
Fix scrolling issue

geido commented:
> > @kgabryje @kasiazjc Thank you for the explanation and I think the new fix for the apply/clear button look better.
> > I found another cosmetic issue of scrolling up the select box, i think sometimes if user scroll up the bar when select box is open, the select box will be scroll up too, also the content of dashboard is scrolled up
> > scrolling.issue.mov
> 
> That's super weird 😕 I can't reproduce this issue locally, but I'm fairly sure that it's not related to my changes. @geido @michael-s-molina Are you guys able to reproduce?

@kgabryje @jinghua-qa There is a PR where that issue is being worked on https://github.com/apache/superset/pull/18102. @pkdotson do you have any update on that issue?

geido commented:
/testenv up

geido reviewed (approved):
LGTM!

jinghua-qa commented:
> > > @kgabryje @kasiazjc Thank you for the explanation and I think the new fix for the apply/clear button look better.
> > > I found another cosmetic issue of scrolling up the select box, i think sometimes if user scroll up the bar when select box is open, the select box will be scroll up too, also the content of dashboard is scrolled up
> > > scrolling.issue.mov
> > 
> > 
> > That's super weird 😕 I can't reproduce this issue locally, but I'm fairly sure that it's not related to my changes. @geido @michael-s-molina Are you guys able to reproduce?
> 
> @kgabryje @jinghua-qa There is a PR where that issue is being worked on #18102. @pkdotson do you have any update on that issue?

I think you are right, i am also able to reproduce it on master. So it not related to this PR

jinghua-qa reviewed (approved):
LGTM

shenrie commented:
@kgabryje Is it at all possible to use CSS to reposition the apply/clear buttons at the top instead of at the bottom, or at least above the set of filters? I have tried changing the flex ordering of the items in the div with the open class, but it doesn't seem to work as I would expect. 

The problem that this change has introduced for us, is that when embedding a SS report in an iframe, the height of the iframe affects the location of those buttons so when the dashboard is large in length and we have increased the height of the iframe to accommodate, the user has to unnecessarily scroll down to get to those apply/clear buttons - which is not only annoying but also not intuitive. Ideally, the button container would be anchored to the last filter field and not the bottom of the filter bar.

shenrie commented:
@kgabryje Nevermind, I figured out a way to do it with some basic CSS and will leave it here for others in case they also don't like the Apply and Clear buttons at the bottom of the filter box.

Just add this to the dashboard's CSS and modify the top position as needed depending on the number of filtering options are available:

```
.open div:nth-child(3) {
  
  position: absolute;
  top: 250px;
}
```
<img width="1359" alt="image" src="https://user-images.githubusercontent.com/768977/174900205-4493becb-53f1-45bc-be16-06cd1d9e3d44.png">

sfirke commented:
@shenrie thanks for posting your workaround, I was in the same situation.  Once I had 3 filters though it didn't work because it moved the buttons box and the 3rd filter.  So I added `>` to make it non-recursive (only finding direct 3rd children) and it worked:
```
.open > div:nth-child(3) {
  
  position: absolute;
  top: 300px;
}
```

jbat commented:
@shenrie  @sfirke 
Have either of you figured out a more flexible design here?
Ideally a way to allow for the addition of Cross Filter to this work around.

If the buttons are fixed, and then multiple Cross Filters are added it means the standard filters slide down below
the fixed position of the buttons resulting in a very non-optimal look.
 
<img width="322" alt="embedded-filters-issue" src="https://github.com/user-attachments/assets/8464fad1-61cf-4ace-a30c-00503a6cfa42">
